File: python_scripts/homer2meme.py
#!/usr/bin/env python3
"""
homer2meme.py - Convert HOMER motif format to MEME motif format

Usage:
    python homer2meme.py -i motifs.homer > motifs.meme
    python homer2meme.py -i motifs.homer.gz > motifs.meme
    python homer2meme.py -i motifs.json -f json > motifs.meme
    python homer2meme.py -i motifs.homer --input-format logodds > motifs.meme
    python homer2meme.py -i motifs.homer --input-format logodds -b 0.2 > motifs.meme
    python homer2meme.py -i motifs.homer --alphabet ACGU > motifs.meme
    cat motifs.homer | python homer2meme.py -i -

Requires: Python 3.8+, no external dependencies
"""

import argparse
import gzip
import sys

from motif_bridge import __version__
from motif_bridge.io import read_homer, read_json, write_meme


def positive_float(value: str) -> float:
    try:
        v = float(value)
    except ValueError as exc:
        raise argparse.ArgumentTypeError(f"Invalid float value: {value}") from exc
    if v <= 0:
        raise argparse.ArgumentTypeError("-a must be > 0.")
    return v


def background_prob(value: str):
    parts = value.split(",")
    values = []
    for part in parts:
        try:
            v = float(part)
        except ValueError as exc:
            raise argparse.ArgumentTypeError(f"Invalid -b value: {value}") from exc
        if not (0 < v <= 1):
            raise argparse.ArgumentTypeError("-b values must be in (0, 1].")
        values.append(v)
    if len(values) > 1 and abs(sum(values) - 1.0) > 1e-3:
        raise argparse.ArgumentTypeError("-b vector must sum to 1.0.")
    return values if len(values) > 1 else values[0]


def positive_int(value: str) -> int:
    try:
        v = int(value)
    except ValueError as exc:
        raise argparse.ArgumentTypeError(f"Invalid integer value: {value}") from exc
    if v <= 0:
        raise argparse.ArgumentTypeError("--nsites must be > 0.")
    return v


def nonnegative_float(value: str) -> float:
    try:
        v = float(value)
    except ValueError as exc:
        raise argparse.ArgumentTypeError(f"Invalid float value: {value}") from exc
    if v < 0:
        raise argparse.ArgumentTypeError("--evalue must be >= 0.")
    return v


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Convert HOMER motif format to MEME format.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=(
            "Examples:\n"
            "  %(prog)s -i results/motifs.homer > raw/motifs.meme\n"
            "  %(prog)s -i results/motifs.homer.gz > raw/motifs.meme\n"
            '  %(prog)s -i results/motifs.homer -e "CTCF/Jaspar"\n'
            "  %(prog)s -i motifs.json -f json > motifs.meme\n"
            "  %(prog)s -i motifs.homer --input-format logodds\n"
            "  %(prog)s -i motifs.homer --input-format logodds -b 0.2\n"
            "  %(prog)s -i motifs.homer --input-format logodds -b 0.29,0.21,0.21,0.29\n"
            "  %(prog)s -i motifs.homer --alphabet ACGU\n"
            "  cat motifs.homer | %(prog)s -i -\n"
        ),
    )
    parser.add_argument(
        "-i",
        metavar="<file>",
        required=True,
        help="Input HOMER format file (or '-' for stdin, supports .gz)",
    )
    parser.add_argument(
        "-e",
        metavar="<string>",
        default="",
        help="Extract only specified motif by id or description",
    )
    parser.add_argument(
        "-a",
        metavar="<float>",
        type=positive_float,
        default=0.01,
        help="Pseudocount for log-odds -> probability (default: 0.01)",
    )
    parser.add_argument(
        "-b",
        "--background",
        metavar="<float[,float...]>",
        type=background_prob,
        default=0.25,
        help="Background probability, either scalar or comma-separated vector (default: 0.25)",
    )
    parser.add_argument(
        "-f",
        "--format",
        choices=["homer", "json"],
        default="homer",
        help="Input format: homer (default) or json",
    )
    parser.add_argument(
        "--input-format",
        choices=["auto", "logodds", "probability"],
        default="auto",
        help="Matrix type: auto (default), logodds, or probability",
    )
    parser.add_argument(
        "--alphabet",
        choices=["ACGT", "ACGU", "PROTEIN"],
        default="ACGT",
        help="Alphabet for HOMER input: ACGT (default), ACGU, or PROTEIN",
    )
    parser.add_argument(
        "--rc",
        action="store_true",
        help="Output the reverse complement of the motif (DNA/RNA only).",
    )
    parser.add_argument(
        "--trim-edges",
        metavar="<float>",
        type=float,
        default=0.0,
        help="Trim edges with information content below threshold.",
    )
    parser.add_argument(
        "--min-ic",
        metavar="<float>",
        type=float,
        default=0.0,
        help="Filter out motifs with total information content below threshold.",
    )
    parser.add_argument(
        "--nsites",
        metavar="<int>",
        type=positive_int,
        default=None,
        help="Override MEME nsites metadata in output.",
    )
    parser.add_argument(
        "--evalue",
        metavar="<float>",
        type=nonnegative_float,
        default=None,
        help="Override MEME E metadata in output.",
    )
    parser.add_argument(
        "--renormalize",
        action="store_true",
        help="Renormalize each row before writing MEME output.",
    )
    parser.add_argument(
        "--strict",
        action="store_true",
        help=(
            "Fail on malformed matrix rows, ambiguous auto-detection, or invalid probability rows."
        ),
    )
    parser.add_argument("--version", action="version", version=f"%(prog)s {__version__}")
    return parser.parse_args()


def open_input(path: str):
    """Open plain, gzip, or stdin input."""
    if path == "-":
        return sys.stdin
    elif path.endswith(".gz"):
        return gzip.open(path, "rt", encoding="utf-8")
    else:
        return open(path, "r", encoding="utf-8")


def process_motifs(motifs, args):
    """Filter and apply operations to motifs."""
    for m in motifs:
        if args.e and m.id != args.e and m.description != args.e:
            continue

        if args.rc:
            try:
                m.reverse_complement()
            except ValueError as exc:
                sys.stderr.write(f"Warning: skipping motif '{m.id}': {exc}\n")
                continue
        if args.trim_edges > 0:
            m.trim_edges(args.trim_edges)
        if not m.matrix:
            continue
        if args.min_ic > 0 and m.total_ic() < args.min_ic:
            continue

        yield m


def main() -> None:
    args = parse_args()
    fh = None
    try:
        fh = open_input(args.i)

        if args.format == "json":
            raw_motifs = read_json(
                fh,
                pseudocount=args.a,
                input_format=args.input_format,
                background=args.background,
                strict=args.strict,
            )
        else:
            raw_motifs = read_homer(
                fh,
                pseudocount=args.a,
                input_format=args.input_format,
                alphabet=args.alphabet,
                background=args.background,
                strict=args.strict,
            )

        processed_motifs = process_motifs(raw_motifs, args)
        write_meme(
            processed_motifs,
            sys.stdout,
            nsites=args.nsites,
            evalue=args.evalue,
            renormalize=args.renormalize,
            background=args.background,
            strict=args.strict,
        )

    except FileNotFoundError:
        sys.exit(f"Error: Cannot open file: {args.i}")
    except ValueError as exc:
        sys.exit(f"Error: {exc}")
    except BrokenPipeError:
        pass
    finally:
        if fh is not None and args.i != "-":
            try:
                fh.close()
            except Exception:
                pass


if __name__ == "__main__":
    main()

File: rust_scripts/src/bin/homer2meme.rs
//! homer2meme - Convert HOMER motif format to MEME motif format
//!
//! Build (Cargo, recommended - includes .gz support):
//!   cd rust_scripts && cargo build --release
//!   # Binaries: rust_scripts/target/release/meme2homer  homer2meme
//!
//! Install to PATH:
//!   cd rust_scripts && cargo install --path .
//!
//! Usage:
//!   homer2meme -i motifs.homer > motifs.meme
//!   homer2meme -i motifs.homer.gz > motifs.meme
//!   homer2meme -i motifs.json -f json > motifs.meme
//!   homer2meme -i motifs.homer --input-format logodds > motifs.meme
//!   cat motifs.homer | homer2meme -i -

use clap::{ArgAction, Parser, ValueEnum};
use flate2::read::MultiGzDecoder;
use motif_bridge::io::{
    read_homer_with_strict, read_json_with_strict, write_meme_with_background, MatrixType,
};
use std::fs::File;
use std::io::{self, BufRead, BufReader, BufWriter};

#[derive(Parser)]
#[command(
    name = "homer2meme",
    about = "Convert HOMER motif format to MEME format.",
    version,
    after_help = "Examples:\n  homer2meme -i motifs.homer > motifs.meme\n  homer2meme -i motifs.homer.gz > motifs.meme\n  homer2meme -i motifs.homer -e \"CTCF/Jaspar\"\n  homer2meme -i motifs.json -f json > motifs.meme\n  homer2meme -i motifs.homer --input-format logodds\n  cat motifs.homer | homer2meme -i -\n"
)]
struct Args {
    /// Input HOMER motif file (.homer, .motif, .json, or .gz), or '-' for stdin
    #[arg(short = 'i', value_name = "<file>")]
    input: String,
    /// Extract only specified motif by id or description
    #[arg(short = 'e', default_value = "")]
    extract: String,
    /// Pseudocount for log-odds -> probability conversion
    #[arg(short = 'a', value_parser = parse_positive, default_value = "0.01")]
    pseudocount: f64,
    /// Background probability scalar or comma-separated vector
    #[arg(short = 'b', long = "background", default_value = "0.25")]
    background: String,
    /// Input format: homer (default) or json
    #[arg(short = 'f', long = "format", value_enum, default_value_t = InputFormat::Homer)]
    input_format: InputFormat,
    /// Matrix type: auto (default), logodds, or probability
    #[arg(long = "input-format", value_enum, default_value_t = MatrixTypeArg::Auto)]
    matrix_type: MatrixTypeArg,
    /// Alphabet for HOMER input
    #[arg(long = "alphabet", value_enum, default_value_t = Alphabet::Acgt)]
    alphabet: Alphabet,
    /// Output the reverse complement of the motif (DNA/RNA only)
    #[arg(long = "rc", action = ArgAction::SetTrue)]
    rc: bool,
    /// Trim edges with information content below threshold
    #[arg(long = "trim-edges", value_parser = parse_nonneg, default_value = "0.0")]
    trim_edges: f64,
    /// Filter out motifs with total information content below threshold
    #[arg(long = "min-ic", value_parser = parse_nonneg, default_value = "0.0")]
    min_ic: f64,
    /// Override MEME nsites metadata in output
    #[arg(long = "nsites", value_parser = parse_positive_usize)]
    nsites: Option<usize>,
    /// Override MEME E metadata in output
    #[arg(long = "evalue", value_parser = parse_nonneg)]
    evalue: Option<f64>,
    /// Renormalize each row before writing MEME output
    #[arg(long = "renormalize", action = ArgAction::SetTrue)]
    renormalize: bool,
    /// Fail on malformed matrix rows, ambiguous auto-detection, or invalid probability rows
    #[arg(long = "strict", action = ArgAction::SetTrue)]
    strict: bool,
}

#[derive(Clone, Copy, ValueEnum)]
enum InputFormat {
    #[value(name = "homer")]
    Homer,
    #[value(name = "json")]
    Json,
}

#[derive(Clone, Copy, ValueEnum)]
enum MatrixTypeArg {
    #[value(name = "auto")]
    Auto,
    #[value(name = "logodds")]
    Logodds,
    #[value(name = "probability")]
    Probability,
}

#[derive(Clone, Copy, ValueEnum)]
enum Alphabet {
    #[value(name = "ACGT")]
    Acgt,
    #[value(name = "ACGU")]
    Acgu,
    #[value(name = "PROTEIN")]
    Protein,
}

impl Alphabet {
    fn as_str(self) -> &'static str {
        match self {
            Alphabet::Acgt => "ACGT",
            Alphabet::Acgu => "ACGU",
            Alphabet::Protein => "PROTEIN",
        }
    }
}

impl MatrixTypeArg {
    fn to_matrix_type(self) -> MatrixType {
        match self {
            MatrixTypeArg::Auto => MatrixType::Auto,
            MatrixTypeArg::Logodds => MatrixType::Logodds,
            MatrixTypeArg::Probability => MatrixType::Probability,
        }
    }
}

fn parse_positive(value: &str) -> Result<f64, String> {
    let v = value
        .parse::<f64>()
        .map_err(|_| format!("invalid -a value: {}", value))?;
    if v <= 0.0 {
        return Err(format!("-a must be > 0, got {}", v));
    }
    Ok(v)
}

fn parse_background(value: &str) -> Result<Vec<f64>, String> {
    let mut values = Vec::new();
    for part in value.split(',') {
        let v = part
            .parse::<f64>()
            .map_err(|_| format!("invalid -b value: {}", value))?;
        if v <= 0.0 || v > 1.0 {
            return Err(format!("-b values must be in (0, 1], got {}", v));
        }
        values.push(v);
    }
    if values.is_empty() {
        return Err("-b must contain at least one value".to_string());
    }
    if values.len() > 1 {
        let total: f64 = values.iter().sum();
        if (total - 1.0).abs() > 1e-3 {
            return Err(format!("-b vector must sum to 1.0, got {:.6}", total));
        }
    }
    Ok(values)
}

fn parse_nonneg(value: &str) -> Result<f64, String> {
    let v = value
        .parse::<f64>()
        .map_err(|_| format!("invalid value: {}", value))?;
    if v < 0.0 {
        return Err(format!("value must be >= 0, got {}", v));
    }
    Ok(v)
}

fn parse_positive_usize(value: &str) -> Result<usize, String> {
    let v = value
        .parse::<usize>()
        .map_err(|_| format!("invalid --nsites value: {}", value))?;
    if v == 0 {
        return Err("--nsites must be > 0".to_string());
    }
    Ok(v)
}

fn run() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();
    let background = parse_background(&args.background)
        .map_err(|e| io::Error::new(io::ErrorKind::InvalidInput, e))?;

    let reader: Box<dyn BufRead> = if args.input == "-" {
        Box::new(BufReader::new(io::stdin().lock()))
    } else if args.input.ends_with(".gz") {
        let file = File::open(&args.input)?;
        Box::new(BufReader::new(MultiGzDecoder::new(file)))
    } else {
        let file = File::open(&args.input)?;
        Box::new(BufReader::new(file))
    };

    let matrix_type = args.matrix_type.to_matrix_type();
    let raw_motifs = match args.input_format {
        InputFormat::Json => read_json_with_strict(
            reader,
            args.pseudocount,
            matrix_type,
            &background,
            args.strict,
        )?,
        InputFormat::Homer => read_homer_with_strict(
            reader,
            args.pseudocount,
            matrix_type,
            args.alphabet.as_str(),
            &background,
            args.strict,
        )?,
    };

    let mut processed_motifs = Vec::new();
    for mut m in raw_motifs {
        if !args.extract.is_empty() && m.id != args.extract && m.description != args.extract {
            continue;
        }

        if args.rc {
            if let Err(e) = m.reverse_complement() {
                eprintln!("Warning: skipping motif '{}': {}", m.id, e);
                continue;
            }
        }
        if args.trim_edges > 0.0 {
            m.trim_edges(args.trim_edges);
        }
        if m.matrix.is_empty() {
            continue;
        }
        if args.min_ic > 0.0 && m.total_ic() < args.min_ic {
            continue;
        }
        processed_motifs.push(m);
    }

    let mut stdout = BufWriter::new(io::stdout().lock());
    let output_background = if background.len() > 1 {
        Some(background.as_slice())
    } else {
        None
    };
    write_meme_with_background(
        &mut stdout,
        &processed_motifs,
        args.nsites,
        args.evalue,
        args.renormalize,
        output_background,
        args.strict,
    )?;

    Ok(())
}

fn main() {
    if let Err(e) = run() {
        eprintln!("Error: {}", e);
        std::process::exit(1);
    }
}
